android/src/main/java/com/oney/WebRTCModule/PermissionUtils.java

```diff
@@ -0,0 +1,133 @@
+     */
+    private static final String GRANT_RESULT = "GRANT_RESULT";
+    private static final String PERMISSION = "PERMISSION";
+    private static final String REQUEST_CODE = "REQUEST_CODE";
+    private static final String RESULT_RECEIVER = "RESULT_RECEIVER";
+
+    /**
+     * Incrementing counter for permissions requests. Each request must have a unique numeric code.
+     */
+    private static int requestCode;
+
+    /**
+     * Requests the given permission. The callback will we called with the requested permission and
+     * the granted result. See: https://developer.android.com/reference/android/content/pm/PackageManager.html#PERMISSION_GRANTED
+     * for the result constants.
+     *
+     * @param context Application context / activity.
+     * @param permission Permission which the application is interested in requesting.
+     * @param callback Callback where the results will be reported.
+     */
+    public static void requestPermission(Context context,
+                                         String permission,
+                                         final Callback callback) {
+        ResultReceiver resultReceiver = new ResultReceiver(new Handler(Looper.getMainLooper())) {
+            @Override
+            protected void onReceiveResult (int resultCode, Bundle resultData) {
+                String permission = resultData.getString(PERMISSION);
+                int grantResult = resultData.getInt(GRANT_RESULT);
+
+                callback.invoke(permission, grantResult);
+            }
+        };
+
+        requestCode++;
+
+        Intent permissionsIntent = new Intent(context, PermissionHelperActivity.class);
+        permissionsIntent.putExtra(RESULT_RECEIVER, resultReceiver);
+        permissionsIntent.putExtra(PERMISSION, permission);
+        permissionsIntent.putExtra(REQUEST_CODE, requestCode);
+        permissionsIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+
+        context.startActivity(permissionsIntent);
+    }
+
+    /**
+     * Helper activity for requesting permissions. Android only allows requesting permissions from
+     * an activity and the result is reported in the <tt>onRequestPermissionsResult</tt> method.
+     * Since this package is a library we create an auxiliary activity and communicate back the
+     * results using a <tt>ResultReceiver</tt>.
+     */
+    public static class PermissionHelperActivity extends Activity {
+        int requestCode;
+        ResultReceiver resultReceiver;
+        String permission;
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void onRequestPermissionsResult(int requestCode,
+                                               String permissions[],
+                                               int[] grantResults) {
+            Bundle resultData = new Bundle();
+            resultData.putString(PERMISSION, permissions[0]);
+            resultData.putInt(GRANT_RESULT, grantResults[0]);
```

Comment:
The cancellation scenario presents an unclear situation. For example:
1. JS calls `getUserMedia`
2. `PermissionUtils` requests a permission i.e. a dialog is presented to the user.
3. The user switches away from the app, possibly on a subject unrelated to the permission request.
4. The user comes back to the application.

At step 4, `onRequestPermissionsResult` will signal a cancellation. But not all permissions will have been reported as granted or denied to `callback` so `getUserMedia` will be stuck waiting for an answer to a question that is no longer posed to the user because `finish()` was invoked (presuming there's no unguarded access to elements at index 0, of course).

One question to be answered is whether `PermissionUtil` should ask for the permission again since technically it was not answered by the user. I guess we can look at what iOS gives as an experience in such a scenario.